### PR TITLE
Update interpreter and test suite for table64

### PR DESCRIPTION
**Pull request by sbc100**

### CODE REVIEW

**Comment by sbc100** [comment]
I didn't end up duplicating all of the table.* tests here but just enough to get a little coverage for each instruction.   Let me know if you think its worth duplicating more.

File: test/core/table.wast [hunk 1]
```diff
@@ -24,17 +24,17 @@
   "size minimum must not be greater than maximum"
 )
 
-(assert_malformed
+(assert_invalid
   (module quote "(table 0x1_0000_0000 funcref)")
-  "i32 constant out of range"
+  "table size must be at most 2^32-1"
```
Comment (rossberg) [inline test/core/table.wast:25]: Should we still be imposing this here? Shouldn't it become an implementation limit now?

File: test/core/table64.wast [hunk 2]
```diff
@@ -0,0 +1,26 @@
+;; Test table section structure
+;; Largely duplicated from table.wast, but with all tables using a 64-bit index.
```
Comment (rossberg) [inline test/core/table64.wast]: Given that all the other table64 tests are in the existing files, I think this should be merged into table.wast as well.

File: interpreter/exec/eval.ml [hunk 3]
```diff
@@ -42,6 +42,11 @@ let numeric_error at = function
   | exn -> raise exn
 
 
+let value_of_index it x =
```
Comment (rossberg) [inline interpreter/exec/eval.ml]: Perhaps move this to value.ml

File: interpreter/runtime/table.mli [hunk 4]
```diff
@@ -16,7 +16,9 @@ exception OutOfMemory
 
 val alloc : table_type -> ref_ -> table (* raises Type, OutOfMemory *)
 val type_of : table -> table_type
+val index_of : table -> index_type
```
Comment (rossberg) [inline interpreter/runtime/table.mli]:
```suggestion
val index_type_of : table -> index_type
```

File: interpreter/exec/eval.ml [hunk 5]
```diff
@@ -241,85 +247,97 @@ let rec step (c : config) : config =
...
 
       | TableSize x, vs ->
-        Num (I32 (Table.size (table frame.inst x))) :: vs, []
+        let tab = table frame.inst x in
+        value_of_index (Table.index_of tab) (Table.size (table frame.inst x)) :: vs, []
 
-      | TableGrow x, Num (I32 delta) :: Ref r :: vs' ->
+      | TableGrow x, Num delta :: Ref r :: vs' ->
         let tab = table frame.inst x in
+        let delta_64 = Table.index_of_num delta in
```
Comment (rossberg) [inline interpreter/exec/eval.ml]: Nit: Perhaps inline this.

File: interpreter/runtime/table.ml [hunk 6]
```diff
@@ -17,47 +17,62 @@ exception OutOfMemory
...
+  | Some m -> I64.le_u min m
 
-let create size r =
-  try Lib.Array32.make size r
+let valid_index it i =
+  match it with
+  | I32IndexType -> I64.le_u i 0xffff_ffffL
+  | I64IndexType -> true
+
+let create size it r =
```
Comment (rossberg) [inline interpreter/runtime/table.ml]: Why do we need it here?

File: interpreter/text/arrange.ml [hunk 7]
```diff
@@ -533,9 +533,9 @@ let func f =
 (* Tables & memories *)
 
 let table off i tab =
-  let {ttype = TableType (lim, t)} = tab.it in
-  Node ("table $" ^ nat (off + i) ^ " " ^ limits nat32 lim,
-    [atom ref_type t]
+  let {ttype = TableType (lim, it, t)} = tab.it in
+  Node ("table $" ^ nat (off + i) ^  " " ^ index_type it ^  " " ^
```
Comment (rossberg) [inline interpreter/text/arrange.ml]:
```suggestion
  Node ("table $" ^ nat (off + i) ^ " " ^ index_type it ^ " " ^
```

**Review by rossberg — COMMENTED** [review]
LGTM, mostly nits, except for the question about the table size limit.

File: interpreter/runtime/table.ml (hunk 6 above)
Comment (sbc100) [reply interpreter/runtime/table.ml]: Oh I guess maybe we don't.  I think I just copied that from the memory.ml.  Fixed.

File: interpreter/exec/eval.ml (hunk 3 above)
Comment (sbc100) [reply interpreter/exec/eval.ml]: I don't see that file.. are you suggesting a new file with this one function?

File: test/core/table64.wast (hunk 2 above)
Comment (sbc100) [reply test/core/table64.wast]: Done.

File: test/core/table.wast (hunk 1 above)
Comment (sbc100) [reply test/core/table.wast:25]: Are you suggesting that we should drop this limit check, even for i32 tables?  Currently this limit only applies to i32 tables.

File: interpreter/exec/eval.ml (hunk 3 above)
Comment (rossberg) [reply interpreter/exec/eval.ml]: Sorry, syntax/values.ml

File: test/core/table.wast (hunk 1 above)
Comment (rossberg) [reply test/core/table.wast:25]: Ah, sorry, no. That makes sense.

File: test/core/table_init.wast [hunk 8]
```diff
@@ -33,9 +36,13 @@
   (func (result i32) (i32.const 8))
   (func (result i32) (i32.const 9))  ;; index 9
   (func (export "test")
-    (table.init $t0 1 (i32.const 7) (i32.const 0) (i32.const 4)))
+    (table.init $t0 2 (i32.const 7) (i32.const 0) (i32.const 4)))
+  (func (export "test-t64")
+    (table.init $t64 2 (i32.const 7) (i32.const 0) (i32.const 4)))
```
Comment (alexcrichton) [inline test/core/table_init.wast:41]: I think this trailing `i32.const` needs to be `i64.const`?

**Review by alexcrichton — COMMENTED** [review]
Poking at this I think there's one issue I flagged below, but mind also adding tests for `table.copy`? That'll likely want to inherit the same logic for `memory.copy` where the src/dst offsets have static types but the length-to-copy is the minimum of the two sizes.

File: test/core/table_init.wast (hunk 8 above)
Comment (sbc100) [reply test/core/table_init.wast:41]: Thanks for catching.. taking a look now.

**Pull request merged**